File: python/gradbench/gradbench/tools/tensorflow/ht.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

# https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/python/modules/TensorflowGraph/TensorflowGraphHand.py

import numpy as np
import tensorflow as tf
from tensorflow.python.framework.ops import disable_eager_execution

disable_eager_execution()  # turn eager execution off

from gradbench import wrap
from gradbench.adbench.ht_data import HandInput
from gradbench.adbench.itest import ITest
from gradbench.tools.tensorflow.ht_objective import (
    ht_objective,
    ht_objective_complicated,
)
from gradbench.tools.tensorflow.utils import flatten, to_tf_tensor


class TensorflowHT(ITest):
    """Test class for HT diferentiation by Tensorflow using computational
    graphs."""

    def prepare(self, input):
        """Prepares calculating. This function must be run before any others."""

        self.complicated = len(input.us) > 0
        self.nrows = 3 * len(input.data.correspondences)

        graph = tf.compat.v1.Graph()
        with graph.as_default():
            if self.complicated:
                self.variables = np.append(input.us.flatten(), input.theta)
                self.params = (
                    input.data.model.bone_count,
                    input.data.model.parents,
                    to_tf_tensor(input.data.model.base_relatives),
                    to_tf_tensor(input.data.model.inverse_base_absolutes),
                    to_tf_tensor(input.data.model.base_positions),
                    tf.transpose(to_tf_tensor(input.data.model.weights)),
                    input.data.model.is_mirrored,
                    to_tf_tensor(input.data.points),
                    input.data.correspondences,
                    input.data.model.triangles,
                )

                self.objective_function = ht_objective_complicated
                self.ncols = len(input.theta) + 2
            else:
                self.variables = input.theta
                self.params = (
                    input.data.model.bone_count,
                    input.data.model.parents,
                    to_tf_tensor(input.data.model.base_relatives),
                    to_tf_tensor(input.data.model.inverse_base_absolutes),
                    to_tf_tensor(input.data.model.base_positions),
                    tf.transpose(to_tf_tensor(input.data.model.weights)),
                    input.data.model.is_mirrored,
                    to_tf_tensor(input.data.points),
                    input.data.correspondences,
                )

                self.objective_function = ht_objective
                self.ncols = len(input.theta)

            self.prepare_operations()

        self.session = tf.compat.v1.Session(graph=graph)
        self.first_running()

        self.objective = None
        self.jacobian = None

    def prepare_operations(self):
        """Prepares computational graph for needed operations."""

        self.variables_holder = tf.compat.v1.placeholder(
            dtype=tf.float64, shape=self.variables.shape
        )

        with tf.GradientTape(persistent=True) as grad_tape:
            grad_tape.watch(self.variables_holder)
            self.objective_operation = self.objective_function(
                self.variables_holder, *self.params
            )

            self.objective_operation = flatten(self.objective_operation)

        self.jacobian_operation = grad_tape.jacobian(
            self.objective_operation, self.variables_holder, experimental_use_pfor=False
        )

        self.feed_dict = {self.variables_holder: self.variables}

    def first_running(self):
        """Performs the first session running."""

        self.session.run(self.objective_operation, feed_dict=self.feed_dict)

        self.session.run(self.jacobian_operation, feed_dict=self.feed_dict)

    def calculate_objective(self):
        self.objective = self.session.run(
            self.objective_operation, feed_dict=self.feed_dict
        )

    def calculate_jacobian(self):
        self.jacobian = self.session.run(
            self.jacobian_operation, feed_dict=self.feed_dict
        )

        if self.complicated:
            # Merging us part of jacobian to two columns.
            # Note: jacobian has the following structure:
            #
            #   [us_part theta_part]
            #
            # where us part is a block diagonal matrix with blocks of
            # size [3, 2]

            us_J = np.concatenate(
                [
                    self.jacobian[3 * i : 3 * i + 3, 2 * i : 2 * i + 2]
                    for i in range(self.nrows // 3)
                ],
                0,
            )

            us_count = 2 * self.nrows // 3
            theta_J = self.jacobian[:, us_count:]

            self.jacobian = np.concatenate((us_J, theta_J), 1)


def prepare_input(input):
    py = TensorflowHT()
    py.prepare(HandInput.from_dict(input))
    return py


def objective_output(output):
    return output.flatten().tolist()


def jacobian_output(output):
    return output.tolist()


@wrap.multiple_runs(pre=prepare_input, post=objective_output)
def objective(py):
    py.calculate_objective()
    return py.objective


@wrap.multiple_runs(pre=prepare_input, post=jacobian_output)
def jacobian(py):
    py.calculate_jacobian()
    return py.jacobian

File: tools/finite/ht.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// Largely derived from https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/cpp/modules/finite/FiniteHand.cpp

#include <algorithm>
#include "gradbench/main.hpp"
#include "gradbench/evals/ht.hpp"
#include "finite.h"

class Jacobian : public Function<ht::Input, ht::JacOutput> {
  bool _complicated = false;
  std::vector<double> _objective;
  std::vector<double> _jacobian_by_us;
  FiniteDifferencesEngine<double> _engine;
public:
  Jacobian(ht::Input& input) :
    Function(input),
    _complicated(input.us.size() != 0),
    _objective(3 * input.data.correspondences.size()),
    _jacobian_by_us(2 * _objective.size())
  {
    _engine.set_max_output_size(_objective.size());
  }

  void compute(ht::JacOutput& output) {
    int err_size = 3 * _input.data.correspondences.size();
    int ncols = (_complicated ? 2 : 0) + _input.theta.size();
    output.jacobian_ncols = ncols;
    output.jacobian_nrows = err_size;
    output.jacobian.resize(err_size * ncols);

    if (_complicated) {
      _engine.finite_differences(1, [&](double* theta_in, double* err) {
        ht::objective(theta_in, _input.us.data(), &_input.data, err);
      }, _input.theta.data(), _input.theta.size(),
        _objective.size(),
        &output.jacobian.data()[6 * _input.data.correspondences.size()]);

      for (unsigned int j = 0; j < _input.us.size() / 2; ++j) {
        _engine.finite_differences(1, [&](double* us_in, double* err) {
          ht::objective(_input.theta.data(), us_in, &_input.data, err);
        }, _input.us.data(), _input.us.size(), j * 2, 2,
          _objective.size(),
          _jacobian_by_us.data());

        for (int k = 0; k < 3; ++k) {
          output.jacobian[j * 3 + k] = _jacobian_by_us[j * 3 + k];
          output.jacobian[j * 3 + k + _objective.size()] = _jacobian_by_us[j * 3 + k + _objective.size()];
        }
      }
    } else {
      _engine.finite_differences(1, [&](double* theta_in, double* err) {
        ht::objective(theta_in, &_input.data, err);
      }, _input.theta.data(), _input.theta.size(), _objective.size(), output.jacobian.data());
    }
  }
};

int main(int argc, char* argv[]) {
  return generic_main(argc, argv, {
      {"objective", function_main<ht::Objective>},
      {"jacobian", function_main<Jacobian>}
    });
}
